validate.go

```diff
@@ -18,34 +19,127 @@ func validate(payload []byte) ([]byte, error) {
+		name := gjson.GetBytes(payload, "request.object.metadata.name").String()
+		namespace := gjson.GetBytes(payload,
+			"request.object.metadata.namespace").String()
 		e.String("name", name)
 		e.String("namespace", namespace)
 	})
 
-	if settings.DeniedNames.Contains(name) {
-		logger.InfoWithFields("rejecting ingress object", func(e onelog.Entry) {
+	for _, volume := range volumes.Array() {
+		if gjson.Get(volume.String(), "hostPath").Exists() {
+			volumeName := gjson.Get(volume.String(), "name").String()
+
+			// obtain volumeMount object matching the 'volume.name'
+			mount := gjson.Get(volumeMounts.String(),
+				fmt.Sprintf("#(name==\"%s\")", volumeName))
+
+			if mount.Exists() {
+				// is a hostpath and is in use, validate against settings
+
+				readOnly := false // volumeMount.readOnly missing means 'false'
+				if gjson.Get(mount.String(), "readOnly").Exists() {
+					readOnly = gjson.Get(mount.String(), "readOnly").Bool()
+				}
+
+				path := gjson.Get(volume.String(), "hostPath.path").String()
+
+				match := false
+				// readOnly attribute of most specific AllowedHostPath takes precendence:
+				previousAllowedHostPath := ""
+				for _, allowedHostPath := range settings.AllowedHostPaths {
+					if hasPathPrefix(path, allowedHostPath.PathPrefix) {
+						// current setting allowedHostPath matches path of volumeMount
+						if hasPathPrefix(allowedHostPath.PathPrefix, previousAllowedHostPath) {
+							// allowedHostPath is more specific (and has precendence over
+							//	past allowedHostPath), or the same path
+							match = true
+							err = validatePath(path, readOnly, allowedHostPath)
+							previousAllowedHostPath = allowedHostPath.PathPrefix
+						}
+					}
+				}
+				if !match {
+					// path didn't match against any PathPrefix in settings
+					err = fmt.Errorf("hostPath '%s' is not in the AllowedHostPaths list",
+						path)
+				}
+				if err != nil {
+					break // stop first for, a path was not allowed against all settings, reject
+				}
+			}
+		}
+	}
+
+	if err != nil {
+		logger.DebugWithFields("rejecting pod object", func(e onelog.Entry) {
+			name := gjson.GetBytes(payload, "request.object.metadata.name").String()
+			namespace := gjson.
+				GetBytes(payload, "request.object.metadata.namespace").String()
 			e.String("name", name)
-			e.String("denied_names", settings.DeniedNames.String())
+			e.String("namespace", namespace)
 		})
-
 		return kubewarden.RejectRequest(
-			kubewarden.Message(
-				fmt.Sprintf("The '%s' name is on the deny list", name)),
+			kubewarden.Message(err.Error()),
 			kubewarden.NoCode)
 	}
 
-	logger.Info("accepting ingress object")
 	return kubewarden.AcceptRequest()
 }
+
+// validatePath validates the path prefix and its readOnly state against the
+// passed hostPath, and returns a matching error if failed.
+func validatePath(path string, readOnly bool, hostPath HostPath) (err error) {
+	if hasPathPrefix(path, hostPath.PathPrefix) {
```

Comment:
I don't think w need this outer `if` condition. The `validatePath` is called only when this condition is verified (see line 86)

Reply:
I think we can't drop this line; line 86 checks for precedence of paths (if the saved last allowed path is less specific than current one), and this line checks the mount against the current settings' prefix.